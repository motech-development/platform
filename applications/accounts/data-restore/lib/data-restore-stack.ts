import { Duration, Stack, StackProps } from 'aws-cdk-lib';
import { Table } from 'aws-cdk-lib/aws-dynamodb';
import { Runtime } from 'aws-cdk-lib/aws-lambda';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { Construct } from 'constructs';

interface IConfig {
  source: string;
  target: string;
}

export class DataRestoreStack extends Stack {
  private readonly source: string;

  private readonly target: string;

  constructor(
    scope: Construct,
    id: string,
    props: StackProps,
    config: IConfig,
  ) {
    super(scope, id, props);

    this.source = config.source;

    this.target = config.target;

    this.restore();
  }

  private getDatabases() {
    const sourceName = this.source;

    const source = Table.fromTableAttributes(this, sourceName, {
      tableName: sourceName,
    });

    const targetName = this.target;

    const target = Table.fromTableAttributes(this, targetName, {
      tableName: targetName,
    });

    return {
      source,
      target,
    };
  }

  private restore() {
    const { source, target } = this.getDatabases();

    const handler = new NodejsFunction(this, 'data-restore', {
      bundling: {
        minify: true,
      },
      entry: './src/data-restore.ts',
      environment: {
        LOG_LEVEL: 'debug',
        SOURCE: source.tableName,
        TARGET: target.tableName,
      },
      functionName: 'data-restore',
      runtime: Runtime.NODEJS_20_X,
      timeout: Duration.minutes(15),
    });

    source.grantReadData(handler);

    target.grantFullAccess(handler);
  }
}
