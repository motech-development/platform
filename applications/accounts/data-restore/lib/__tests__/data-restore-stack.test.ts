import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { DataRestoreStack } from '../data-restore-stack';

describe('DomainStack', () => {
  let app: App;

  beforeEach(() => {
    app = new App();
  });

  it('should synthesize as expected', () => {
    const stack = new DataRestoreStack(
      app,
      'TestStack',
      {},
      {
        source: 'source-table',
        target: 'target-table',
      },
    );

    const template = Template.fromStack(stack);

    expect(template).toMatchSnapshot();
  });
});
