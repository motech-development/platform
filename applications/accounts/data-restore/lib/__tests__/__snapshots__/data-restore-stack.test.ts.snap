// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataRestoreStack should synthesize as expected 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "datarestoreD8FFD2B4": {
      "DependsOn": [
        "datarestoreServiceRoleDefaultPolicyD37817A2",
        "datarestoreServiceRoleB13C83ED",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "bc25fd9bbc3bf0c0c4a5371ab0c2687f6476d9faa8f48bc9ac7d64ba192ba17e.zip",
        },
        "Environment": {
          "Variables": {
            "LOG_LEVEL": "debug",
            "SOURCE": "source-table",
            "TARGET": "target-table",
          },
        },
        "FunctionName": "data-restore",
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "datarestoreServiceRoleB13C83ED",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "datarestoreServiceRoleB13C83ED": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "datarestoreServiceRoleDefaultPolicyD37817A2": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":dynamodb:",
                      {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":table/source-table",
                    ],
                  ],
                },
                {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
            {
              "Action": "dynamodb:*",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":dynamodb:",
                      {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":table/target-table",
                    ],
                  ],
                },
                {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "datarestoreServiceRoleDefaultPolicyD37817A2",
        "Roles": [
          {
            "Ref": "datarestoreServiceRoleB13C83ED",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
